cdap-standalone/src/main/java/co/cask/cdap/StandaloneMain.java

```diff
@@ -326,23 +327,38 @@ public void shutDown() {
       }
 
       if (zkClient != null) {
-        zkClient.startAndWait();
+        zkClient.stopAndWait();
       }
+    } catch (Throwable e) {
+      halt = true;
+      LOG.error("Exception during shutdown", e);
+    } finally {
+      stopExternalProcesses();
+      cleanupTempDir();
+    }
+
+    // We can't do much but exit. Because there was an exception, some non-daemon threads may still be running.
+    // Therefore System.exit() won't do it, we need to force a halt.
+    if (halt) {
+      Runtime.getRuntime().halt(1);
+    }
+  }
 
-      if (kafkaProcessExecutor != null) {
+  private void stopExternalProcesses() {
+    if (kafkaProcessExecutor != null) {
+      try {
         kafkaProcessExecutor.stopAndWait();
+      } catch (UncheckedExecutionException ex) {
+        LOG.warn("Exception during Kafka process shutdown.", ex);
```

Comment:
Getting pid is non-trivial since we don't have any nice abstraction in Java Process. I can add - ``Process might still be running``.